File: day_19/p1.py
def parse(fname: str):
    lines = []
    with open(fname, 'r') as f:
        lines = [line for line in f.read().splitlines() if line]
    
    towels = lines[0].split(', ')
    designs = lines[1:]
    
    return towels, designs

def dp(target, towels, memo):
    if target in memo.keys():
        return memo[target]

    if len(target) == 0:
        return True
    
    for t in towels:
        if target.startswith(t):
            can_compose = dp(target[len(t):], towels, memo)
            if can_compose:
                memo[target] = True
                return True

    memo[target] = False
    return memo[target]

def compute(towels, designs):
    count = 0
    for d in designs:
        memo = {}
        if dp(d, towels, memo):
            count += 1
    
    return count

def main():
    towels, designs = parse("input.txt")
    print(compute(towels, designs))


if __name__ == "__main__":
    main()

File: day_19/p1.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>

std::pair<strings, strings> Parse(const std::string& fname)
{
    strings towels {};
    strings designs {};
    auto lines = pypp::splitFileLines(fname);

    for (int i { 0 }; i < lines.size(); ++i) {
        if (i == 0)
            towels = pypp::split(lines[i], ',');
        else
            designs.push_back(lines[i]);
    }

    for (auto& t : towels)
        t = pypp::strip(t, [](unsigned char ch) { return std::isspace(ch); });

    return { towels, designs };
}

bool dp(const std::string& target, const strings& towels, std::unordered_map<std::string, bool>& memo)
{
    if (memo.contains(target))
        return memo[target];

    if (target.length() == 0)
        return true;

    for (auto t : towels) {
        if (target.find(t) == 0) {
            if (dp(target.substr(t.length()), towels, memo)) {
                memo[target] = true;
                return true;
            }
        }
    }

    memo[target] = false;
    return false;
}

int Compute(const strings& towels, const strings& designs)
{
    auto count { 0 };

    for (auto d : designs) {
        std::unordered_map<std::string, bool> memo {};
        if (dp(d, towels, memo))
            count++;
    }

    return count;
}

int main()
{
    const auto [towels, designs] = Parse("input.txt");
    std::cout << Compute(towels, designs) << std::endl;

    return 0;
}
